distributed/worker.py

```diff
@@ -1563,146 +1638,392 @@ def add_task(
-                    # Either `fetch` -> `waiting` or `flight` -> `waiting`
-                    self.log.append((ts.key, "re-adding key, new TaskState"))
-                    self.transition(ts, "waiting", runspec=runspec)
+        self.log.append((key, "compute-task", stimulus_id))
+        if key in self.tasks:
+            logger.debug(
+                "Asked to compute an already known task %s",
+                {"task": self.tasks[key], "stimulus_id": stimulus_id},
+            )
+            ts = self.tasks[key]
+        else:
+            self.tasks[key] = ts = TaskState(key)
+
+        ts.runspec = SerializedTask(function, args, kwargs, task)
+
+        if priority is not None:
+            priority = tuple(priority) + (self.generation,)
+            self.generation -= 1
+
+        if actor:
+            self.actors[ts.key] = None
+
+        ts.exception = None
+        ts.traceback = None
+        ts.scheduler_holds_ref = True
+        ts.priority = priority
+        ts.duration = duration
+        if resource_restrictions:
+            ts.resource_restrictions = resource_restrictions
+        ts.annotations = annotations
+
+        recommendations = {}
+        scheduler_msgs = []
+        for dependency, _ in who_has.items():
+            recs, s_msgs = self.register_data_task_internal(
+                key=dependency,
+                stimulus_id=stimulus_id,
+                priority=priority,
+            )
+            recommendations.update(recs)
+            scheduler_msgs.extend(s_msgs)
+            dep_ts = self.tasks[dependency]
+
+            # link up to child / parents
+            ts.dependencies.add(dep_ts)
+            dep_ts.dependents.add(ts)
+
+        if ts.state in ("memory", "error"):
+            # TODO: Factor this out to adhere to the recommendation/msg system
+            self.send_task_state_to_scheduler(ts)
+        elif ts.state in ("ready", "executing", "waiting"):
+            pass
+        elif ts.state in ("released", "fetch", "flight"):
+            recommendations[ts] = "waiting"
+        else:
+            raise RuntimeError(f"Unexpected task state encountered {ts} {stimulus_id}")
+
+        for msg in scheduler_msgs:
+            self.batched_stream.send(msg)
+        self.transitions(recommendations, stimulus_id=stimulus_id)
+
+        # We received new info, that's great but not related to the compute-task
+        # instruction
+        self.update_who_has(who_has=who_has)
+        if nbytes is not None:
+            for key, value in nbytes.items():
+                self.tasks[key].nbytes = value
+
+    def transition_released_fetch__recs(self, ts):
+        if self.validate:
+            assert ts.state == "released"
+            assert ts.runspec is None
+
+        for w in ts.who_has:
+            self.pending_data_per_worker[w].append(ts.key)
+        ts.state = "fetch"
+        heapq.heappush(self.data_needed, (ts.priority, ts.key))
+        return {}, []
```

Comment:
If we want to keep track on scheduler side of "who is fetching what" we could emit a message in here